idx/memory/memory.go

```diff
@@ -1095,49 +1139,96 @@ func (m *MemoryIdx) delete(orgId int, n *Node, deleteEmptyParents, deleteChildre
 // delete series from the index if they have not been seen since "oldest"
 func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]idx.Archive, error) {
 	oldestUnix := oldest.Unix()
-	var pruned []idx.Archive
-	pre := time.Now()
-	orgs := []int{orgId}
+	orgs := make(map[int]struct{})
 	if orgId == -1 {
 		log.Info("memory-idx: pruning stale metricDefs across all orgs")
 		m.RLock()
-		orgs = make([]int, len(m.Tree))
-		i := 0
 		for org := range m.Tree {
-			orgs[i] = org
-			i++
+			orgs[org] = struct{}{}
+		}
+		if TagSupport {
+			for org := range m.tags {
+				orgs[org] = struct{}{}
+			}
 		}
 		m.RUnlock()
+	} else {
+		orgs[orgId] = struct{}{}
 	}
-	for _, org := range orgs {
-		m.RLock()
-		tree, ok := m.Tree[org]
-		if !ok {
-			m.RUnlock()
-			continue
+
+	var pruned []idx.Archive
+	toPruneUntagged := make(map[int]map[string]struct{}, len(orgs))
+	toPruneTagged := make(map[int]IdSet, len(orgs))
+	for org := range orgs {
+		toPruneTagged[org] = make(IdSet)
+		toPruneUntagged[org] = make(map[string]struct{})
+	}
+	pre := time.Now()
+
+	m.RLock()
+DEFS:
+	for _, def := range m.DefById {
+		if _, ok := orgs[def.OrgId]; !ok {
+			continue DEFS
 		}
 
-		var toPrune []string
+		if def.LastUpdate >= oldestUnix {
+			continue DEFS
+		}
 
-		for _, n := range tree.Items {
-			if !n.Leaf() {
-				continue
+		if len(def.Tags) == 0 {
+			tree, ok := m.Tree[def.OrgId]
+			if !ok {
+				continue DEFS
 			}
-			staleCount := 0
+
+			n, ok := tree.Items[def.Name]
+			if !ok || !n.Leaf() {
+				continue DEFS
+			}
+
 			for _, id := range n.Defs {
-				if m.DefById[id].LastUpdate < oldestUnix {
-					staleCount++
+				if m.DefById[id].LastUpdate >= oldestUnix {
+					continue DEFS
 				}
 			}
-			if staleCount == len(n.Defs) {
-				//we need to delete this node.
-				toPrune = append(toPrune, n.Path)
+
+			toPruneUntagged[def.OrgId][n.Path] = struct{}{}
+		} else {
+			id, err := idx.NewMetricIDFromString(def.Id)
+			if err != nil {
+				log.Error(3, "memory-idx: corrupt index. ID format %s seems to be invalid", def.Id)
+				continue DEFS
 			}
+
+			toPruneTagged[def.OrgId][id] = struct{}{}
```

Comment:
This doesn't take into account there being multiple metricDefs with the same tagset due to the interval of the metric changing (or schema.MetricDefinition.SetId() changing).   When there are multiple metricDefs with the same tagSet we need to keep all of them as it means that series is still in use.

This is why when deleting from the tree, we check to make sure the leaf node has no other metricDefs that are still being updated.

An easy way to solve this would be to add a `DefByTagSet map[string][]*idx.Archive` to the MemoryIdx struct.
where the key of the map is just the series' tags sorted and concatenated together.  Then when pruning, we use the old tree search for non-tagged series, and also iterate over the DefByTagSet   map to find series where all metricDefs are stale.

Reply:
right, i wrongly assumed that unlike the tree index there's going to be only one metric def per combo. I forgot about the interval. will fix that

Reply:
Reading your proposal with the `DefByTagSet`, I can already hear @Dieterbe telling us that there are too many strings in the indexes and we don't want more because of gc overhead for references. 
Unfortunately the only good alternative I can think of would involve creating a `TagQuery` for each of the series that are candidates for pruning, but I think that's probably going to cause too much over head.

How about this though:
Instead of using those string indexes for `DefByTagSet` as you describe, we could use a key that is an md5 hash (of static size, not involving references). Then, when the pruning wants to check if for a pruning-candidate there are other series with the same name + tags combo, it could get its md5 hash and lookup the list of other defs with the same md5 hash from `DefByTagSet`. If there are hash collisions it is not a problem because the pruning process could simply ignore those with a matching MD5 but where name + tags don't match the pruning candidate it is currently checking for. 
That way we get all the good things: 1) not many references and 2) it's fast

Or am I making that unnecessarily complicated?

Reply:
> Or am I making that unnecessarily complicated?

Yes.   Just use a string.  Strings are only problematic when you start manipulating them as changing a string requires an allocation and copy.
Even if you used a hash of the tagset, that hash value would need to be a []byte (there would be way too many collisions for an integer hash like Murmur or FNV-1a).  A []byte has more overheads then a string. https://syslog.ravelin.com/byte-vs-string-in-go-d645b67ca7ff

Reply:
Ok, no problem. 
Btw I didn't mean to use `[]byte`, since an md5 hash is always 128 bits we could use something like a struct of two `uint64` to store it. But I'll use strings anyway

Reply:
> Strings are only problematic when you start manipulating them as changing a string requires an allocation and copy.

yes and no. avoiding allocations is a good thing when we can, especially in hot code, because then you can see speedups and GC is triggered less frequently. (most of our allocations currently happens in kafka-mdm data decoding, every single metric packet ingested triggers a bunch of string allocations)
However the problem replay was referring to is amount of live pointers. the more live pointers (pointers that stick around for minutes, hours, days or weeks) you have, the more the GC has to work each time it runs. and this has an averse effect on performance, slowing down our goroutines to an extent sometimes significantly more than what the stop-the-world reported would lead one to believe. (see #172)

this is why storing stuff in pointerless structures such as integers or fixed-size arrays is beneficial, in general. maybe not so much if the flipside is that you have to calculate checksums. i haven't gone through this particular code yet.

Reply:
I think this should solve the described problem then: https://github.com/grafana/metrictank/pull/806/commits/cc0c3d10c46cbc9abb8c1c7c113577caa1861c4a
If in the future we want to get rid of the strings as map index and replace it with something else, like statically sized MD5s or whatever, then that would be really easy.

Reply:
@woodsaj that would be ready for review again if you have time